File: 백준/Gold/13146. 같은 수로 만들기 2/같은 수로 만들기 2.py
import sys

input = sys.stdin.readline


n = int(input())
a = [int(input()) for i in range(n)]


result = 0

stack = []
for i in a:
    diff = 0
    while stack and stack[-1] <= i:
        diff = max(diff, i-stack.pop())
    result+=diff
    stack.append(i)

if len(stack)>1:
    result += stack[0]-stack[-1]

print(result)

File: 백준/Gold/13146. 같은 수로 만들기 2/같은 수로 만들기 2.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();
        List<Integer> a = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            a.add(sc.nextInt());
        }

        long result = 0;
        Deque<Integer> stack = new ArrayDeque<>();

        for (int i : a) {
            int diff = 0;
            while (!stack.isEmpty() && stack.peekLast() <= i) {
                diff = Math.max(diff, i - stack.pollLast());
            }
            result += diff;
            stack.addLast(i);
        }

        if (stack.size() > 1) {
            result += stack.peekFirst() - stack.peekLast();
        }

        System.out.println(result);
    }
}
